bouncer/src/repo/manipulator/modeloptimizer/repo_optimizer_multipart.cpp

```diff
@@ -399,75 +917,234 @@ bool MultipartOptimizer::isTransparent(
+	// original Id so processSupermeshMaterials can find the node to
+	// copy (when necessary).
+
+	MaterialUUIDMap stashIdToOriginalId;
+	for (const auto mapping : materialMap)
+	{
+		stashIdToOriginalId[mapping.second] = mapping.first;
+	}
+
+	for (auto supermesh : supermeshes)
+	{
+		// First create a new entry parented to the scene root. This will be the final 
+		// object.
+
+		auto sMeshWithParent = supermesh->cloneAndAddParent({ rootID });
+		supermesh->swap(sMeshWithParent);
+		mergedMeshes.insert(supermesh);
+
+		// Next, create the re-mapped material nodes on demand, ensuring the new nodes
+		// store the new mesh in their parents array.
+
+		processSupermeshMaterials(scene, supermesh, mergedMeshesMaterials, stashIdToOriginalId);
+	}
+
+	return true;
+}
+
+void MultipartOptimizer::processSupermeshMaterials(
+	const repo::core::model::RepoScene* scene,
+	const repo::core::model::MeshNode* supermeshNode,
+	MaterialNodes& mergedMaterialNodes,
+	MaterialUUIDMap& stashIdToOriginalIdMap
+)
+{
+	std::set<repo::lib::RepoUUID> supermeshStashMaterials;
+	for (const auto& map : supermeshNode->getMeshMapping())
+	{
+		supermeshStashMaterials.insert(map.material_id);
+	}
+
+	// The id of this final supermesh MeshNode. This is used to set the parents 
+	// array of the material nodes.
+
+	repo::lib::RepoUUID sMeshSharedID = supermeshNode->getSharedID();
 
-			auto matIt = matNodes.find(matID.second);
-			if (matIt == matNodes.end())
+	for (const auto stashMaterialId : supermeshStashMaterials)
+	{
+		// Has another supermesh created a Node for this duplicated material already?
+		auto existing = mergedMaterialNodes.find(stashMaterialId);
+		if (existing == mergedMaterialNodes.end())
+		{
+			// No, so we must clone the material here.
+			auto originalNode = scene->getNodeByUniqueID(defaultGraph, stashIdToOriginalIdMap[stashMaterialId]);
+			if (originalNode)
 			{
-				//This material hasn't beenn copied yet.
-				//clone and wipe the parent entries, insert new parents
-				auto matNode = scene->getNodeByUniqueID(defaultGraph, matID.first);
-				if (matNode)
-				{
-					repo::core::model::RepoNode clonedMat = repo::core::model::RepoNode(matNode->removeField(REPO_NODE_LABEL_PARENTS));
-					repo::core::model::RepoBSONBuilder builder;
-					builder.append(REPO_NODE_LABEL_ID, matID.second);
-					auto changeBSON = builder.obj();
-					clonedMat = clonedMat.cloneAndAddFields(&changeBSON, false);
-					clonedMat = clonedMat.cloneAndAddParent({ sMeshSharedID });
-					matNodes[matID.second] = new repo::core::model::MaterialNode(clonedMat);
-				}
+				repo::core::model::RepoNode clonedMat = repo::core::model::RepoNode(originalNode->removeField(REPO_NODE_LABEL_PARENTS));
+				repo::core::model::RepoBSONBuilder builder;
+				builder.append(REPO_NODE_LABEL_ID, stashMaterialId);
+				auto changeBSON = builder.obj();
+				clonedMat = clonedMat.cloneAndAddFields(&changeBSON, false);
+				clonedMat = clonedMat.cloneAndAddParent({ sMeshSharedID });
```

Comment:
It's better to wait till you've found all the meshes that needs to link to all the parents and add them in one go

In the past we've had this call really slowing down performance when we had to call `cloneAndAddParent` multiple times

Reply:
I.e. i think it's better if you pass all the supermeshes down and only process this once

Reply:
I've refactored this so one function handles all the supermeshes/cloned materials at the end of generateMultipartScene.